openmdao/core/driver.py

```diff
@@ -588,6 +596,60 @@ def run(self, problem):
 
         self.recorders.record(system, metadata)
 
+    def calc_gradient(self, indep_list, unknown_list, mode='auto',
+                      return_format='array'):
+        """ Returns the scaled gradient for the system that is slotted in
+        self.root, scaled by all scalers that were specified when the desvars
+        and constraints were added.
+
+        Args
+        ----
+        indep_list : list of strings
+            List of independent variable names that derivatives are to
+            be calculated with respect to. All params must have a IndepVarComp.
+
+        unknown_list : list of strings
+            List of output or state names that derivatives are to
+            be calculated for. All must be valid unknowns in OpenMDAO.
+
+        mode : string, optional
+            Deriviative direction, can be 'fwd', 'rev', 'fd', or 'auto'.
+            Default is 'auto', which uses mode specified on the linear solver
+            in root.
+
+        return_format : string, optional
+            Format for the derivatives, can be 'array' or 'dict'.
+
+        Returns
+        -------
+        ndarray or dict
+            Jacobian of unknowns with respect to params.
+        """
+
+        dv_conversions = {}
```

Comment:
You might just want to calculate dv_conversions and cn_conversions for all _desvars and all _objs and _cons in driver setup and cache them.  It shouldn't matter if those dicts end up with entries in them that you don't need, since you never iterate over them.